components/script/devtools.rs

```diff
@@ -397,6 +397,65 @@ pub(crate) fn handle_get_layout(
         .unwrap();
 }
 
+pub(crate) fn handle_get_xpath(
+    documents: &DocumentCollection,
+    pipeline: PipelineId,
+    node_id: String,
+    reply: IpcSender<String>,
+) {
+    let Some(node) = find_node_by_unique_id(documents, pipeline, &node_id) else {
+        return reply.send(Default::default()).unwrap();
+    };
+
+    let selector = node
+        .inclusive_ancestors(ShadowIncluding::Yes)
+        .filter_map(|ancestor| {
+            let Some(element) = ancestor.downcast::<Element>() else {
+                // TODO: figure out how to handle shadow roots here
```

Comment:
Is this because the xpath spec predates shadowroots, or a technical limitation in Servo?

Reply:
Its because there's no way for an XPath expression to move into a shadow root.